[TSan] Add positive test for TSan + Dispatch on Linux

yln opened the pull request:
1) Enable tests that use `import Dispatch` on Linux. Add substitution
       `%import-libdispatch` that needs to be used for all cross-platform
       tests (i.e., tests that are intended to be run on other platforms
       than Darwin) that do `import Dispatch` or enable thread sanitizer.
    
2) Make sure as many existing Dispatch and TSan tests as possible run on
       Linux. Mark tests that would require substantial work with
       `UNSUPPORTED: OS=linux-gnu`.
    
3) Add integration-style Swift test that shows that TSan finds a simple
       race when using `Dispatch.async` incorrectly. A more complete test
       suite for TSan's libdispatch support lives on the LLVM/compiler-rt
       side.
    
rdar://problem/49177535

yln committed:
[TSan] Add positive test for TSan + Dispatch on Linux

1) Enable tests that use `import Dispatch` on Linux. Add substitution
   `%import-libdispatch` that needs to be used for all cross-platform
   tests (i.e., tests that are intended to be run on other platforms
   than Darwin) that do `import Dispatch` or enable thread sanitizer.

2) Make sure as many existing Dispatch and TSan tests as possible run on
   Linux. Mark tests that would require substantial work with
   `UNSUPPORTED: OS=linux-gnu`.

3) Add integration-style Swift test that shows that TSan finds a simple
   race when using `Dispatch.async` incorrectly. A more complete test
   suite for TSan's libdispatch support lives on the LLVM/compiler-rt
   side.

rdar://problem/49177535

yln commented on test/stdlib/Dispatch.swift:
```diff
@@ -1,6 +1,7 @@
 // RUN: %target-run-simple-swift
 // REQUIRES: executable_test
 // REQUIRES: libdispatch
+// XFAIL: OS=linux-gnu
```
```
/home/yln/work/swift-source/swift/test/stdlib/Dispatch.swift:14:42: error: type 'DispatchSource' has no member 'ProcessEvent'
  expectEqual(2147483648, DispatchSource.ProcessEvent.exit.rawValue)
                          ~~~~~~~~~~~~~~ ^~~~~~~~~~~~
/home/yln/work/swift-source/swift/test/stdlib/Dispatch.swift:87:13: error: use of unresolved identifier 'mach_timebase_info_data_t'
        var info = mach_timebase_info_data_t(numer: 1, denom: 1)
                   ^~~~~~~~~~~~~~~~~~~~~~~~~
/home/yln/work/swift-source/swift/test/stdlib/Dispatch.swift:88:2: error: use of unresolved identifier 'mach_timebase_info'
        mach_timebase_info(&info)
        ^~~~~~~~~~~~~~~~~~
/home/yln/work/swift-source/swift/test/stdlib/Dispatch.swift:95:46: error: use of unresolved identifier 'NSEC_PER_SEC'
        time = DispatchTime(uptimeNanoseconds: 15 * NSEC_PER_SEC)
                                                    ^~~~~~~~~~~~
/home/yln/work/swift-source/swift/test/stdlib/Dispatch.swift:96:43: error: use of unresolved identifier 'NSEC_PER_SEC'
        expectEqual(time.uptimeNanoseconds, 15 * NSEC_PER_SEC)
                                                 ^~~~~~~~~~~~
/home/yln/work/swift-source/swift/test/stdlib/Dispatch.swift:148:14: error: use of unresolved identifier 'NSEC_PER_SEC'
        expectEqual(NSEC_PER_SEC, diffNanos)
                    ^~~~~~~~~~~~
/home/yln/work/swift-source/swift/test/stdlib/Dispatch.swift:153:66: error: use of unresolved identifier 'O_RDONLY'
        let chan = DispatchIO(type: .random, path: "_REL_PATH_", oflag: O_RDONLY, mode: 0, queue: q, cleanupHandler: { (error) in })
                                                                        ^~~~~~~~
/home/yln/work/swift-source/swift/test/stdlib/Dispatch.swift:200:36: error: argument type 'DispatchTimeInterval' does not conform to expected type 'Equatable'
                expectEqual(DispatchTimeInterval.seconds(i), DispatchTimeInterval.milliseconds(i * 1000))
                                                 ^
/home/yln/work/swift-source/swift/test/stdlib/Dispatch.swift:201:36: error: argument type 'DispatchTimeInterval' does not conform to expected type 'Equatable'
                expectEqual(DispatchTimeInterval.milliseconds(i), DispatchTimeInterval.microseconds(i * 1000))
                                                 ^
/home/yln/work/swift-source/swift/test/stdlib/Dispatch.swift:202:36: error: argument type 'DispatchTimeInterval' does not conform to expected type 'Equatable'
                expectEqual(DispatchTimeInterval.microseconds(i), DispatchTimeInterval.nanoseconds(i * 1000))
                                                 ^
/home/yln/work/swift-source/swift/test/stdlib/Dispatch.swift:228:46: error: binary operator '!=' cannot be applied to two 'DispatchTimeInterval' operands
        expectTrue(DispatchTimeInterval.seconds(10) != DispatchTimeInterval.never);
                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/yln/work/swift-source/swift/test/stdlib/Dispatch.swift:228:46: note: binary operator '!=' cannot be synthesized for enums with associated values
        expectTrue(DispatchTimeInterval.seconds(10) != DispatchTimeInterval.never);
                                                    ^
/home/yln/work/swift-source/swift/test/stdlib/Dispatch.swift:229:40: error: binary operator '!=' cannot be applied to two 'DispatchTimeInterval' operands
        expectTrue(DispatchTimeInterval.never != DispatchTimeInterval.seconds(10));
                   ~~~~~~~~~~~~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/yln/work/swift-source/swift/test/stdlib/Dispatch.swift:229:40: note: binary operator '!=' cannot be synthesized for enums with associated values
        expectTrue(DispatchTimeInterval.never != DispatchTimeInterval.seconds(10));
                                              ^
```

compnerd replied:
Sounds like `-DDEPLOYMENT_TARGET_LINUX` may be missing?

yln replied:
I searched for `DEPLOYMENT_TARGET_LINUX` and it seems to be a swift-corelibs-foundation option. Currently I am building simply via `./swift/utils/build-script  --release-debuginfo  --libdispatch --install-libdispatch`, e.g. only worrying about libdispatch at the moment.

Do you have an idea how to make `import Dispatch` work in the tests without copying things into the <swift-build> directory?

jrose-apple replied:
Add search paths using lit, I think.

compnerd replied:
For the link, you will need to add `-L` for the search paths.  For the runtime, you will need to add `LD_LIBRARY_PATH` on Linux, `DYLD_LIBRARY_PATH` on macOS, and `PATH` on Windows.

yln commented on test/stdlib/DispatchTypes.swift:
```diff
@@ -1,6 +1,7 @@
 // RUN: %target-swift-frontend -typecheck %s
 
 // REQUIRES: libdispatch
+// XFAIL: OS=linux-gnu
```
```
/home/yln/work/swift-source/swift/test/stdlib/DispatchTypes.swift:27:45: error: use of unresolved identifier 'mach_port_t'
_ = DispatchSource.makeMachSendSource(port: mach_port_t(0), eventMask: [])
                                            ^~~~~~~~~~~
/home/yln/work/swift-source/swift/test/stdlib/DispatchTypes.swift:28:48: error: use of unresolved identifier 'mach_port_t'
_ = DispatchSource.makeMachReceiveSource(port: mach_port_t(0))
                                               ^~~~~~~~~~~
/home/yln/work/swift-source/swift/test/stdlib/DispatchTypes.swift:29:20: error: type 'DispatchSource' has no member 'makeMemoryPressureSource'
_ = DispatchSource.makeMemoryPressureSource(eventMask: [])
    ~~~~~~~~~~~~~~ ^~~~~~~~~~~~~~~~~~~~~~~~
/home/yln/work/swift-source/swift/test/stdlib/DispatchTypes.swift:30:20: error: type 'DispatchSource' has no member 'makeProcessSource'; did you mean 'makeWriteSource'?
_ = DispatchSource.makeProcessSource(identifier: 0, eventMask: [])
    ~~~~~~~~~~~~~~ ^~~~~~~~~~~~~~~~~
                   makeWriteSource
Dispatch.DispatchSource:32:23: note: 'makeWriteSource' declared here
    public class func makeWriteSource(fileDescriptor: Int32, queue: DispatchQueue? = nil) -> DispatchSourceWrite
                      ^
/home/yln/work/swift-source/swift/test/stdlib/DispatchTypes.swift:32:45: error: use of unresolved identifier 'SIGINT'
_ = DispatchSource.makeSignalSource(signal: SIGINT)
                                            ^~~~~~
/home/yln/work/swift-source/swift/test/stdlib/DispatchTypes.swift:34:20: error: type 'DispatchSource' has no member 'makeFileSystemObjectSource'
_ = DispatchSource.makeFileSystemObjectSource(fileDescriptor: 0, eventMask: [])
    ~~~~~~~~~~~~~~ ^~~~~~~~~~~~~~~~~~~~~~~~~~
```

compnerd replied:
Sounds like `-DDEPLOYMENT_TARGET_LINUX` may be missing?

compnerd commented on test/ClangImporter/Dispatch_test.swift:
```diff
@@ -1,6 +1,7 @@
 // RUN: %target-typecheck-verify-swift
 
 // REQUIRES: libdispatch
+// XFAIL: OS=linux-gnu
```
Sounds like `-DDEPLOYMENT_TARGET_LINUX` may be missing?

jrose-apple commented on test/Sanitizers/tsan-libdispatch.swift:
```diff
@@ -0,0 +1,36 @@
+
+var racy = 1
+
+for _ in 1...1000 {
```
Feel free to add a substitution for it.

yln commented:
@swift-ci Please test

compnerd commented on test/lit.cfg:
```diff
@@ -905,6 +909,16 @@ elif run_os in ['linux-gnu', 'linux-gnueabihf', 'freebsd', 'windows-cygnus', 'wi
+
+    config.libdispatch_build_dir = make_path(config.swift_obj_root, os.pardir,
+        'libdispatch%s' % config.variant_suffix, 'src')
+    if os.path.exists(config.libdispatch_build_dir):
```
I think that you should make this more resilient and ensure that dispatch, swiftDispatch, and the swift module are available before declaring the feature as being available.

yln replied:
Good point. I will try to make this more resilient.

compnerd commented on test/lit.cfg:
```diff
@@ -905,6 +909,16 @@ elif run_os in ['linux-gnu', 'linux-gnueabihf', 'freebsd', 'windows-cygnus', 'wi
     config.target_runtime = "native"
     config.target_swift_autolink_extract = inferSwiftBinary("swift-autolink-extract")
+
+    config.libdispatch_build_dir = make_path(config.swift_obj_root, os.pardir,
```
This is really unfortunate.  I know that I do not build the same way that the build-script sets things up (even ignoring that the paths are too long that way for Windows).  @davezarzycki may also be building differently.  Can you add an option to CMake that is wired down to the configuration instead?

yln replied:
I am not sure which parts exactly I have to touch for this. Can you point me to an existing variable/configuration option where this is done?

compnerd replied:
I recently wired libxml2 down.  Look at the CMakeList.txt in the test subdirectory.

yln replied:
I am looking at [1] but I am still unsure how to tackle this. Note that the libdispatch that we are referencing here is not built as part of the Swift cmake build, but separately through build-script.
[1] https://github.com/apple/swift/commit/a37a18f16f7211238a50f8606e76dbaee9b60712

yln replied:
We are now wiring the path through CMake!

compnerd commented on test/lit.cfg:
```diff
@@ -1214,7 +1228,7 @@ runtime_libs = {
 }
 
-if run_ptrsize != "32":
+if run_ptrsize != "32" and 'libdispatch' in config.available_features:
```
I think that it is nicer if you wrote this as `run_ptrsize == 64`

compnerd commented on test/stdlib/DispatchData.swift:
```diff
@@ -1,6 +1,6 @@
 // RUN: %empty-directory(%t)
-// RUN: %target-build-swift -swift-version 4 %s -o %t/a.out-4 && %target-codesign %t/a.out-4 && %target-run %t/a.out-4
-// RUN: %target-build-swift -swift-version 4.2 %s -o %t/a.out-4.2 && %target-codesign %t/a.out-4.2 && %target-run %t/a.out-4.2
+// RUN: %target-build-swift -swift-version 4   %s %import-libdispatch -o %t/a.out-4   && %target-codesign %t/a.out-4   && %target-run %t/a.out-4
```
Why the whitespace changes?

compnerd commented on validation-test/Sanitizers/tsan-inout.swift:
```diff
@@ -2,18 +2,23 @@
+#elseif canImport(Glibc)
+  import Glibc
+#else
+#error("Unsupported platform")
```
This doesn't account for Windows

yln replied:
What is the general state of TSan for Swift on Windows? I was unaware that TSan tests ran at all under Windows.

compnerd replied:
Well, all the tests run on Windows, and this is changing the fact that the test should fail to should succeed, but it wont actually do so on Windows.  Marking the TSAN tests as XFAIL on `OS=windows-msvc` is reasonable.

yln replied:
I *think* that `// REQUIRES: tsan_runtime` makes this test not run on Windows (since we don't produce a TSan runtime there). If we start to support TSan on Windows, we have to adapt this test, but I consider this out of scope for this PR.

compnerd commented on validation-test/Sanitizers/tsan-inout.swift:
```diff
@@ -23,8 +28,14 @@ var gInThread2: () -> () = { }
+  var thread1: pthread_t = 0
+  var thread2: pthread_t = 0
+  var t : pthread_t = 0
+#endif
```
Windows doesn't have pthreads

yln replied:
This test didn't run on Windows before and will not run after my change.

compnerd commented on validation-test/Sanitizers/tsan-inout.swift:
```diff
@@ -41,8 +52,13 @@ func testRace(name: String, thread inThread1: @escaping () -> (), thread inThrea
+#else
+  _ = pthread_join(thread1, nil)
+  _ = pthread_join(thread2, nil)
+#endif
```
Same

compnerd reviewed (dismissed):
This is most certainly going to break windows.  There is a community CI up now, you should be able to do a one-off run on there.  Can you add notes to the tests that you marked as unsupported on Linux as to why they are unsupported?

yln commented:
@swift-ci Please test

yln commented:
@swift-ci Please test

yln commented:
> This is most certainly going to break windows. There is a community CI up now, you should be able to do a one-off run on there. Can you add notes to the tests that you marked as unsupported on Linux as to why they are unsupported?

Ah, I missed that before. Can you give/point me to the instructions for doing this?

yln commented:
@swift-ci Please test

yln commented:
@swift-ci Please test

yln commented on test/Sanitizers/tsan-norace-block-release.swift:
```diff
@@ -1,26 +1,31 @@
+#elseif canImport(Glibc)
+  import Glibc
+#else
+#error("Unsupported platform")
```
This test didn't run on Windows before and will not run after my change.

yln commented on test/Sanitizers/tsan-norace-deinit-run-time.swift:
```diff
@@ -1,20 +1,26 @@
+#elseif canImport(Glibc)
+  import Glibc
+#else
+#error("Unsupported platform")
```
This test didn't run on Windows before and will not run after my change.

yln commented:
@swift-ci Please test

yln commented:
@swift-ci Please test

yln commented:
@gottesmm @jrose-apple
The previous CI run succeeded. Assuming the current one does too, is this good to go?

davezarzycki commented:
Since @compnerd mentioned me in this pull request, I just did a clean build of this change using a unified (not `build-script`) setup (Red Hat Fedora 30), and `ninja check-swift-validation` passes. That being said, I'm not trying to test TSan.

yln commented:
@swift-ci Please test

yln commented:
@swift-ci Please test

compnerd commented on test/IRGen/tsan-attributes.swift:
```diff
@@ -2,9 +2,8 @@
-// (We do not test the simulators here.)
-// REQUIRES: CPU=x86_64, OS=macosx
+// TSan is only supported on 64 bit.
+// REQUIRES: PTRSIZE=64
```
Sure, I can try do a run.  BTW, @shahmishal should be able to help you get a run; there is the community CI doing windows builds, so it should be possible to do a one off.

yln replied:
Okay, I contacted Mishal. Waiting for instructions from him.

compnerd commented:
I just did a run of this on Windows and only the following test fails:

Swift(windows-x86_64) :: IRGen/tsan_coroutines.swift

compnerd commented on test/IRGen/tsan_coroutines.swift:
```diff
@@ -23,7 +22,7 @@ extension Foobar {
 
     // We used to crash emitting the subscript function.
-    // CHECK: define swiftcc { i8*, %T15tsan_coroutines1CC* } @"$s15tsan_coroutines6FoobarVyAA1CCAC5IndexVcir"
+    // CHECK: define{{( protected)?}} swiftcc { i8*, %T15tsan_coroutines1CC* } @"$s15tsan_coroutines6FoobarVyAA1CCAC5IndexVcir"
```
This needs a check for `dllexport` on Windows.

yln replied:
Fixed.

yln commented:
@swift-ci Please test

yln commented:
@swift-ci Please test

yln commented:
@swift-ci Please test

jrose-apple commented on test/Sanitizers/tsan-libdispatch.swift:
```diff
@@ -0,0 +1,43 @@
+// RUN: not env %env-TSAN_OPTIONS=abort_on_error=0 %target-run %t_tsan-binary 2>&1 | %FileCheck %s
+// REQUIRES: executable_test
+// REQUIRES: tsan_runtime
+// UNSUPPORTED: OS=tvos
```
What's wrong with tvOS but not watchOS or iOS?

yln replied:
I included it because the other TSan tests, e.g., tsan.swift, tsan-emptyarraystorage.swift, ..., had it.

jrose-apple replied:
Hm, okay.

jrose-apple commented on test/lit.cfg:
```diff
@@ -905,6 +909,21 @@ elif run_os in ['linux-gnu', 'linux-gnueabihf', 'freebsd', 'windows-cygnus', 'wi
+        source_dir = make_path(config.swift_src_root, os.pardir, 'swift-corelibs-libdispatch')
+        swift_module_dir = make_path(libdispatch_build_dir, 'swift')
+        config.import_libdispatch = ('-I %s -I %s -L %s'
+            % (source_dir, swift_module_dir, libdispatch_build_dir))
```
Nitpick: since Python local variables aren't scoped, can you stick `libdispatch_` on the front of `source_dir` and `swift_module_dir` too?

yln replied:
I will do that. My logic was: it's inside the conditional (which is about libdispatch) so I thought it's redundant.

jrose-apple replied:
Not in Python. :-(

jrose-apple commented on test/lit.cfg:
```diff
@@ -1214,7 +1233,7 @@ runtime_libs = {
 }
 
-if run_ptrsize != "32":
+if run_ptrsize == '64' and 'libdispatch' in config.available_features:
```
This still seems wacky. Why rule out TSan-without-libdispatch?

yln replied:
TSan now requires libdispatch, i.e., the TSan runtime has a dependency on it. On MacOS we always have at least the system-provided libdispatch. On Linux, we need to make sure it's available.

In addition, we need to make sure that libdispatch header/swiftmodule/library for linking is available when the test itself uses it. This is what `%import-libdispatch` attempts to do.

jrose-apple commented on test/lit.cfg:
```diff
@@ -1312,14 +1331,15 @@ if os.path.exists(static_libswiftCore_path):
 # default Swift tests to use the just-built libraries
 target_stdlib_path = platform_module_dir
 if not kIsWindows:
+        libdispatch_path = getattr(config, 'libdispatch_build_dir', '')
```
Inconsistent indentation here (although really it's the other way around; @compnerd accidentally snuck some tabs into this file).

yln replied:
Fixed.

yln replied:
Actually, I can't just fix the added line:
```
  File "/home/yln/work/swift-source/swift/test/lit.cfg", line 1335
    if 'use_os_stdlib' not in lit_config.params:
    ^
IndentationError: unexpected indent
```
We can reformat the whole block/section, but I would like to do this in a separate commit.

jrose-apple replied:
My point is that you should use tabs for now, sincee Saleem's code uses tabs.

jrose-apple reviewed (commented):
Sorry to delay on this. This doesn't have to block merging, but I don't love `%import-libdispatch`. What do you all think about adding these search paths unconditionally on Linux?

yln commented:
> ... I don't love `%import-libdispatch`.

I feel the same. The reason for it being like this is that it affects a number of substitutions, e.g., `%target-swift-emit-*`, `%target-swift-frontend`, `%target-run`, `%target-swiftc_driver` and the current approach meant a smaller, more manageable change.

yln commented:
@swift-ci Please test

yln commented:
@swift-ci Please test

yln commented:
Force push to squash commit with uninteresting changes:
1) Align indentation with surrounding code/use tabs.
2) Rename variables.

@swift-ci Please test

yln commented:
@swift-ci Please test

yln committed:
Wire libdispatch build path through CMake

Wire libdispatch build path through CMake `build-script-impl ->
lit.site.cfg.in -> lit.cfg` instead of computing it in lit.cfg.

yln commented:
@swift-ci Please test

yln commented:
@swift-ci Please test

yln commented:
@compnerd @gottesmm @jrose-apple 
I hope I have sufficiently addressed your feedback.

The Swift test suite passes. One lldb test timed out. Can we land this?
```
Timed Out Tests (1):
    lldb-Suite :: lang/swift/stepping/TestSwiftStepping.py
```

yln commented:
@swift-ci Please test OS X platform

yln commented:
@swift-ci Please test

yln commented:
Rebased on top of master.
@swift-ci Please test Linux platform

yln commented:
@compnerd Can you take a final quick look at this?

gottesmm commented:
LGTM if @compnerd is cool. I just pinged him.
